### DISCUSSION TITLE
Smallest subnormal warning for numpy cause numerical issues?

### DISCUSSION THREAD

**Discussion by Andeloth** [NanoComp/meep#2406] [General]
I've been getting this error consistently in meep for the last year
```
/fslhome/andeloth/miniconda3/envs/pmp/lib/python3.10/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.  setattr(self, word, getattr(machar, word).flat[0])
```
and I found a blog post that explains the source of this warning: https://moyix.blogspot.com/2022/09/someones-been-messing-with-my-subnormals.html

Would meep have issues if small subnormals are set to 0 in floating point arithmetic? If there is someone here with computer engineering/computational math background who knows more about this subject then I would love to hear your thoughts on this.

**Comment by Andeloth** [comment]
So after a quick Wikipedia search I found that apparently what `--ffast-math` (the optimization tag in the compilation for the packages listed in the blog post) does is it sets all numbers with smaller exponent than $10^{-95}$ for floats and $10^{-383}$ for doubles -- these are called "subnormal" numbers -- to 0. I don't think I'll ever run any simulations with numbers smaller than that, so I don't think the error will affect my simulations, but when you're running parallel meep the warnings times the number of processes on the output log are sure annoying :D!

**Comment by stevengj** [comment]
Meep currently sets the CPU to flush subnormals to zero (at least on x86/x86_64), because otherwise we get a severe performance penalty at the beginning of simulations when fields are exponentially small.  See #1708

I guess we could try harder to manually set subnormal currents to zero instead, or initialize the fields to random tiny numbers, or ….

(It's pretty rare for any numerical calculation to require nonzero treatment of subnormals.)

**Reply by Andeloth** [reply]
Thanks for linking the issue Steven! Subnormal number treatment is fascinating, even if it’s convenient to flush them and never deal with them instead.

**Reply by oskooi** [reply]
We may want to consider enabling the subnormals feature as a compile-time option. This would leave it up to users to decide whether they want to make use of this feature or not.

**Reply by stevengj** [reply]
(It doesn't need to be a compile-time option, it could be a runtime option like an environment variable.)
